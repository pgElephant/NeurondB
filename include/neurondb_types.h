/*-------------------------------------------------------------------------
 *
 * neurondb_types.h
 *		Core data type definitions for NeuronDB
 *
 * Defines enterprise-grade data types including vectorp (packed SIMD),
 * vecmap (sparse high-dimensional), rtext (retrievable text with tokens),
 * and vgraph (compact graph storage).
 *
 * Copyright (c) 2024-2025, NeuronDB Development Group
 *
 * IDENTIFICATION
 *	  include/neurondb_types.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef NEURONDB_TYPES_H
#define NEURONDB_TYPES_H

#include "postgres.h"

/*
 * vectorp: Packed SIMD vector with metadata
 * - Dimension fingerprint for validation
 * - Endianness guard for portability
 * - Version tag for schema evolution
 */
typedef struct VectorPacked
{
	int32		vl_len_;			/* varlena header */
	uint32		fingerprint;		/* CRC32 of dimensions for validation */
	uint16		version;			/* Schema version */
	uint16		dim;				/* Number of dimensions */
	uint8		endian_guard;		/* 0x01 for little, 0x10 for big */
	uint8		flags;				/* Reserved for future use */
	uint16		unused;				/* Alignment padding */
	float4		data[FLEXIBLE_ARRAY_MEMBER];
} VectorPacked;

/*
 * vecmap: Sparse high-dimensional map
 * - Stores only non-zero values
 * - Mixed float and int keys
 * - Efficient for very high dimensions (>10K)
 */
typedef struct VectorMap
{
	int32		vl_len_;
	int32		total_dim;			/* Total dimensionality */
	int32		nnz;				/* Number of non-zero entries */
	/* Followed by parallel arrays: int32 indices[], float4 values[] */
} VectorMap;

/*
 * rtext: Retrievable text with token metadata
 * - Token offsets for highlighting
 * - Section IDs for structured documents
 * - Language tag for multilingual support
 */
typedef struct RetrievableText
{
	int32		vl_len_;
	uint32		text_len;			/* Length of text in bytes */
	uint16		num_tokens;			/* Number of tokens */
	uint8		lang_tag;			/* Language code (ISO 639-1) */
	uint8		flags;				/* Encoding, case sensitivity, etc */
	/*
	 * Layout:
	 * - char text[text_len]
	 * - uint32 token_offsets[num_tokens]
	 * - uint16 section_ids[num_tokens]
	 */
} RetrievableText;

/*
 * vgraph: Compact graph storage
 * - Node IDs as int64
 * - Typed edges with labels
 * - Adjacency list format
 */
typedef struct VectorGraph
{
	int32		vl_len_;
	int32		num_nodes;
	int32		num_edges;
	uint16		edge_types;			/* Number of edge type labels */
	uint16		unused;
	/*
	 * Layout:
	 * - int64 node_ids[num_nodes]
	 * - Edge edges[num_edges] (src, dst, type, weight)
	 * - char type_labels[][16]
	 */
} VectorGraph;

typedef struct GraphEdge
{
	int32		src_idx;			/* Index into node_ids */
	int32		dst_idx;
	uint16		edge_type;
	uint16		unused;
	float4		weight;
} GraphEdge;

#define VECTORP_SIZE(dim)		(offsetof(VectorPacked, data) + sizeof(float4) * (dim))
#define VECMAP_INDICES(vm)		((int32 *) (((char *)(vm)) + sizeof(VectorMap)))
#define VECMAP_VALUES(vm)		((float4 *) (VECMAP_INDICES(vm) + (vm)->nnz))

#define RTEXT_DATA(rt)			((char *) (((char *)(rt)) + sizeof(RetrievableText)))
#define RTEXT_OFFSETS(rt)		((uint32 *) (RTEXT_DATA(rt) + (rt)->text_len))
#define RTEXT_SECTIONS(rt)		((uint16 *) (RTEXT_OFFSETS(rt) + (rt)->num_tokens))

#define VGRAPH_NODES(vg)		((int64 *) (((char *)(vg)) + sizeof(VectorGraph)))
#define VGRAPH_EDGES(vg)		((GraphEdge *) (VGRAPH_NODES(vg) + (vg)->num_nodes))

#endif /* NEURONDB_TYPES_H */

