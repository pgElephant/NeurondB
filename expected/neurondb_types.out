-- NeurondB Type System Tests
CREATE EXTENSION IF NOT EXISTS neurondb;
-- Test vector creation and I/O
SELECT '[1,2,3]'::vector;
 vector  
---------
 [1,2,3]
(1 row)

SELECT '{1.5, 2.5, 3.5}'::vector;
     vector     
----------------
 [1.5,2.5,3.5]
(1 row)

SELECT '[-1, 0, 1]'::vector;
  vector   
-----------
 [-1,0,1]
(1 row)

-- Test vector dimensions
SELECT vector_dims('[1,2,3,4,5]'::vector);
 vector_dims 
-------------
           5
(1 row)

-- Test vector norms
SELECT vector_norm('[3,4]'::vector);  -- Should be 5.0
 vector_norm 
-------------
           5
(1 row)

SELECT vector_norm('[1,1,1,1]'::vector);  -- Should be 2.0
 vector_norm 
-------------
           2
(1 row)

-- Test vector arithmetic
SELECT '[1,2,3]'::vector + '[4,5,6]'::vector;
 ?column? 
----------
 [5,7,9]
(1 row)

SELECT '[10,20,30]'::vector - '[1,2,3]'::vector;
  ?column?   
-------------
 [9,18,27]
(1 row)

SELECT '[1,2,3]'::vector * 2.0;
 ?column? 
----------
 [2,4,6]
(1 row)

-- Test normalization
SELECT vector_normalize('[3,4]'::vector);
 vector_normalize 
------------------
 [0.6,0.8]
(1 row)

-- Test concatenation
SELECT vector_concat('[1,2]'::vector, '[3,4]'::vector);
 vector_concat 
---------------
 [1,2,3,4]
(1 row)

-- Test array conversion
SELECT vector_to_array('[1,2,3]'::vector);
 vector_to_array 
-----------------
 {1,2,3}
(1 row)

SELECT array_to_vector(ARRAY[1.0, 2.0, 3.0]);
 array_to_vector 
-----------------
 [1,2,3]
(1 row)

-- Test error handling
\set ON_ERROR_STOP off
-- Dimension mismatch
SELECT '[1,2]'::vector + '[1,2,3]'::vector;
ERROR:  vector dimensions must match
-- Invalid input
SELECT '[1,2,a]'::vector;
ERROR:  invalid input syntax for type vector: "[1,2,a]"
SELECT '[]'::vector;
ERROR:  vector must have at least 1 dimension
\set ON_ERROR_STOP on
-- Test quantization
SELECT pg_column_size('[1,2,3,4,5]'::vector) as original_size;
 original_size 
---------------
            28
(1 row)

SELECT pg_column_size(vector_to_int8('[1,2,3,4,5]'::vector)) as int8_size;
 int8_size 
-----------
        13
(1 row)

SELECT pg_column_size(vector_to_binary('[1,2,3,4,5]'::vector)) as binary_size;
 binary_size 
-------------
          11
(1 row)

-- Test binary operations
WITH binary_vecs AS (
    SELECT 
        vector_to_binary('[1, -1, 1, -1]'::vector) as v1,
        vector_to_binary('[1, 1, -1, -1]'::vector) as v2
)
SELECT binary_hamming_distance(v1, v2) FROM binary_vecs;
 binary_hamming_distance 
-------------------------
                       2
(1 row)

-- Performance test: create table with vectors
CREATE TEMP TABLE test_vectors (
    id serial PRIMARY KEY,
    vec vector(128)
);
-- Insert random-ish vectors
INSERT INTO test_vectors (vec)
SELECT ('[' || string_agg((random() * 2 - 1)::text, ',') || ']')::vector
FROM generate_series(1, 128) dim
CROSS JOIN generate_series(1, 1000) row
GROUP BY row;
SELECT COUNT(*) FROM test_vectors;
 count 
-------
  1000
(1 row)

SELECT AVG(vector_norm(vec)) FROM test_vectors;
        avg         
--------------------
 6.5283203125000000
(1 row)

-- Cleanup
DROP TABLE test_vectors;

